### PR TITLE
Module View Command Handler

### PR DESCRIPTION
**Pull request by jpogran**
Adds a handler to return a list of modules in the current workspace to display client side.

Completes: https://github.com/hashicorp/vscode-terraform/pull/746

### CODE REVIEW

File: internal/langserver/handlers/command/module_list.go [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+)
+
+const moduleListVersion = 0
+
+type moduleListResponse struct {
+	FormatVersion int               `json:"v"`
+	Modules       []terraformModule `json:"modules"`
+}
+
+type terraformModule struct {
```
Comment (radeksimko) [inline internal/langserver/handlers/command/module_list.go:23]:
```suggestion
const moduleCallsVersion = 0

type moduleCallsResponse struct {
	FormatVersion int               `json:"v"`
	ModuleCalls       []moduleCall `json:"module_calls"`
}

type moduleCall struct {
```

I think this better represents what the command returns as per https://www.terraform.io/docs/language/modules/syntax.html#calling-a-child-module

File: internal/langserver/handlers/command/module_list.go [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+
+	list := parseModuleRecords(found.ModManifest.Records)
+
+	return moduleListResponse{
+		Modules:       list,
+		FormatVersion: moduleListVersion,
+	}, nil
+}
+
+func parseModuleRecords(records []datadir.ModuleRecord) []terraformModule {
```
Comment (radeksimko) [inline internal/langserver/handlers/command/module_list.go:65]: I think it would be right to add a unit test to cover this fairly complex function.

The input should be relatively easy to mock as part of a test.

File: internal/langserver/handlers/command/module_list.go [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+	modules := make(map[string]terraformModule)
+	for _, manifest := range records {
+		if manifest.Dir == "." {
```
Comment (radeksimko) [inline internal/langserver/handlers/command/module_list.go:68]:
```suggestion
		if manifest.IsRoot() {
```
I _think_ this looks less magical here and you may not even need the comment below at that point.

File: internal/langserver/handlers/command/module_list.go [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+		// in the currecnt workspace by finding a period in the moduleName
+		// is it better to look at SourceAddr and compare?
+		if strings.Contains(manifest.Key, ".") {
+			v := strings.Split(manifest.Key, ".")
+			moduleName = v[0]
+			subModuleName = v[1]
+		}
+
+		// github modules are put in `files` directory, but Manifest Record does not
+		// record the `name` in Key it puts it in SourceAddr
```
Comment (radeksimko) [inline internal/langserver/handlers/command/module_list.go:87]: I think there's some misunderstanding as to what `Key` or `SourceAddr` are for.

`Key` is just locally-relevant reference name within the config (also known as the first and only label of the `module` block), `SourceAddr` is basically 1-1 copy of the raw `source` field within the `module` block.

For example:

```hcl
module "web_server" {
  source = "github.com/hashicorp/two-tier-infra"
}
```
this would be parsed as
 - `Key: "web_server"`
 - `SourceAddr: "github.com/hashicorp/two-tier-infra"`

Both are important and relevant, but not interchangeable.

File: internal/langserver/handlers/command/module_list.go [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+		if strings.HasPrefix(manifest.Key, "files") {
+			moduleName = manifest.SourceAddr
+		}
+
+		// build what we know
+		moduleInfo := terraformModule{
+			Name:    moduleName,
+			Path:    strings.TrimSuffix(manifest.SourceAddr, "/"),
```
Comment (radeksimko) [inline internal/langserver/handlers/command/module_list.go:95]: I think we're mixing two different things here.

As mentioned, `manifest.SourceAddr` is just the "human-typed" address, e.g. `github.com/hashicorp/example` or `hashicorp/consul/aws` (Registry address), it's not strictly speaking a path, but it's probably useful to display in the UI. I guess we can pass this as a separate field `SourceAddr` here?

`manifest.Path` is where the module was actually installed on the filesystem. That is probably _not_ something you would display in the UI, but it may be useful if we want to give users the option to "open installed module" for any local modules? Maybe we can omit `Path` initially.

File: internal/langserver/handlers/command/module_list.go [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+	}
+
+	sort.SliceStable(list, func(i, j int) bool {
+		return list[i].Name < list[j].Name
+	})
+
+	return list
+}
+
+func getModuleType(manifest datadir.ModuleRecord) string {
```
Comment (radeksimko) [inline internal/langserver/handlers/command/module_list.go:129]: It may be worth decoupling this into `internal/terraform/datadir/module_types.go` and relatedly turn the known types into constants.

These can be declared on `*ModuleRecord`.

I think that longer-term we might need to extract it out of the repo entirely, similar to https://github.com/hashicorp/terraform-registry-address but for now keeping it in a separate package should suffice. I can consider the extraction as part of https://github.com/hashicorp/terraform-ls/issues/599 and/or https://github.com/hashicorp/terraform-ls/issues/598

File: internal/langserver/handlers/execute_command.go [hunk 7]
```diff
@@ -12,10 +12,11 @@ import (
...
 var handlers = cmd.Handlers{
-	cmd.Name("rootmodules"):        command.ModulesHandler,
-	cmd.Name("module.callers"):     command.ModuleCallersHandler,
-	cmd.Name("terraform.init"):     command.TerraformInitHandler,
-	cmd.Name("terraform.validate"): command.TerraformValidateHandler,
+	cmd.Name("rootmodules"):          command.ModulesHandler,
+	cmd.Name("module.callers"):       command.ModuleCallersHandler,
+	cmd.Name("terraform.init"):       command.TerraformInitHandler,
+	cmd.Name("terraform.validate"):   command.TerraformValidateHandler,
+	cmd.Name("terraform.modulelist"): command.ModuleListHandler,
```
Comment (radeksimko) [inline internal/langserver/handlers/execute_command.go:19]:
```suggestion
	cmd.Name("module.calls"): command.ModuleCallsHandler,
```
We only prefix the actual commands with `terraform`.

File: internal/langserver/handlers/command/module_list.go [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+	if strings.HasPrefix(manifest.Dir, "modules") {
+		return "local"
+	} else if strings.HasPrefix(manifest.Key, "files") {
+		return "github"
+	} else {
+		return "tfregistry"
```
Comment (radeksimko) [inline internal/langserver/handlers/command/module_list.go:135]: I'm not sure we can make assumptions like above based on the `Dir` or `Key`. I think ideally the only input for detection should be just `manifest.SourceAddr`.

We may be able to reuse some of the logic here https://github.com/hashicorp/terraform/blob/332ea1f233891e45bd274e714ab8e4e11c79dc95/internal/addrs/module_source.go#L191-L268 and relevant tests https://github.com/hashicorp/terraform/blob/332ea1f233891e45bd274e714ab8e4e11c79dc95/internal/addrs/module_source_test.go

or alternatively/additionally the logic in go-getter detectors, which core uses [internally](https://github.com/hashicorp/terraform/blob/332ea1f233891e45bd274e714ab8e4e11c79dc95/internal/getmodules/getter.go#L31-L50):
 - https://github.com/hashicorp/go-getter/blob/main/detect_github.go
 - https://github.com/hashicorp/go-getter/blob/main/detect_git.go
 - https://github.com/hashicorp/go-getter/blob/main/detect_bitbucket.go
 - https://github.com/hashicorp/go-getter/blob/main/detect_gcs.go
 - https://github.com/hashicorp/go-getter/blob/main/detect_s3.go

**Review by radeksimko — COMMENTED** [review]
This is a good start and works for some simple examples.

Aside from my in-line comments/suggestions we'll also need an E2E test for the new command.

Do let me know if you wish to talk through any particular comment, I'm happy to do it more synchronously via Zoom.

File: internal/langserver/handlers/command/module_list.go (hunk 4 above)
Comment (jpogran) [reply internal/langserver/handlers/command/module_list.go]: Thanks for the clarification, this makes much more sense.

File: internal/langserver/handlers/command/module_list.go (hunk 6 above)
Comment (jpogran) [reply internal/langserver/handlers/command/module_list.go]: Moved to new file `internal/terraform/datadir/module_types.go` in this iteration

File: internal/langserver/handlers/command/module_list.go (hunk 8 above)
Comment (jpogran) [reply internal/langserver/handlers/command/module_list.go]: Used `go-getter` and `terraform-registry-address` to parse source addresses.

File: internal/langserver/handlers/command/module_list.go (hunk 5 above)
Comment (jpogran) [reply internal/langserver/handlers/command/module_list.go]: I renamed `path` to `source` to more accurately reflect the idea I was going for in the VS Code UI. Going to omit `manifest.Path` for now, as a TODO for later when we get some user feedback

File: internal/langserver/handlers/command/module_list.go (hunk 2 above)
Comment (jpogran) [reply internal/langserver/handlers/command/module_list.go]: Added tests in `module_calls_test.go`

File: internal/terraform/datadir/module_types.go [hunk 9]
```diff
@@ -0,0 +1,36 @@
+package datadir
+
+import (
+	"github.com/hashicorp/go-getter"
```
Comment (radeksimko) [inline internal/terraform/datadir/module_types.go:4]: I appreciate the convenience aspect of leveraging `go-getter` here directly, but it's worth noting that Terraform unfortunately doesn't use these detectors _directly_ and so its understanding of addresses may differ from go-getter for various reasons, some of which are documented [here](https://github.com/hashicorp/terraform/blob/51b0aee36cc2145f45f5b04051a01eb6eb7be8bf/internal/getmodules/package.go#L7-L36) and [here](https://github.com/hashicorp/terraform/blob/332ea1f233891e45bd274e714ab8e4e11c79dc95/internal/getmodules/getter.go#L13-L28).

All that said it shouldn't differ too much and we only use the detection to improve UX, so the cost of getting it wrong is relatively low and I think it's totally fine to use it for now.

It is likely though that in the long run we will maintain some custom logic which more accurately reflects TF Core implementation and doesn't have any side effects such as making network calls.

TL;DR Ignore the above, the use of go-getter is 👌🏻  in this PR 😄

File: internal/terraform/datadir/module_types.go [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+func (r *ModuleRecord) GetModuleType() ModuleType {
+	if _, err := tfregistry.ParseRawProviderSourceString(r.SourceAddr); err == nil {
+		return TFREGISTRY
+	}
+	if _, ok, _ := new(getter.GitHubDetector).Detect(r.SourceAddr, ""); ok {
+		return GITHUB
+	}
+	if _, ok, _ := new(getter.GitLabDetector).Detect(r.SourceAddr, ""); ok {
+		return GITLAB
+	}
```
Comment (radeksimko) [inline internal/terraform/datadir/module_types.go:27]: Terraform actually doesn't treat GitLab any differently - at least it doesn't use the GitLab detector at all:

https://github.com/hashicorp/terraform/blob/332ea1f233891e45bd274e714ab8e4e11c79dc95/internal/getmodules/getter.go#L30-L52

```suggestion
```

File: internal/terraform/datadir/module_types.go [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+	if _, ok, _ := new(getter.GitDetector).Detect(r.SourceAddr, ""); ok {
+		return GITHUB
```
Comment (radeksimko) [inline internal/terraform/datadir/module_types.go:29]:
```suggestion
		return GIT
```
Typo? 👀

File: internal/terraform/datadir/module_types.go [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+	}
+	if _, ok, _ := new(getter.FileDetector).Detect(r.SourceAddr, ""); ok {
```
Comment (radeksimko) [inline internal/terraform/datadir/module_types.go:31]: Terraform itself doesn't actually use this particular detector - it instead implements its own in https://github.com/hashicorp/terraform/blob/332ea1f233891e45bd274e714ab8e4e11c79dc95/internal/getmodules/file_detector.go#L9-L31 but similar to overall use of go-getter I think the difference is minor enough that we can use it here initially. 👌🏻

File: internal/langserver/handlers/command/module_calls_test.go [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+)
+
+func Test_parseModuleRecords(t *testing.T) {
+	type args struct {
+		records []datadir.ModuleRecord
+	}
+	tests := []struct {
+		name string
+		args args
+		want []moduleCall
```
Comment (radeksimko) [inline internal/langserver/handlers/command/module_calls_test.go:18]:
```suggestion
	tests := []struct {
		name string
		records []datadir.ModuleRecord
		want []moduleCall
```

Nitpick, but I think there is no need for that extra type definition?

**Review by radeksimko — COMMENTED** [review]
Thank you for making these updates, I think it's pretty close to a merge-able state!

I left you some more in-line comments, majority of which are relatively minor/stylistic.

As always - do let me know if you wish to discuss any particular comment/feedback more synchronously via Zoom, I'm happy to do it!

File: internal/langserver/handlers/command/module_calls_test.go (hunk 13 above)
Comment (jpogran) [reply internal/langserver/handlers/command/module_calls_test.go:18]: This is from the neat Go extension generated test cases, and works well with multiple parameter functions but not so well with single parameters.

File: internal/terraform/datadir/module_types.go (hunk 11 above)
Comment (jpogran) [reply internal/terraform/datadir/module_types.go:29]: This was a placeholder until we could decide on icon for git repo based modules. We have landed on `git-branch`, so we can now correctly report the type

File: internal/terraform/datadir/module_types.go (hunk 9 above)
Comment (jpogran) [reply internal/terraform/datadir/module_types.go:4]: 👍🏼 Like you say, we aren't interested in validity here, we're interested in figuring out whether we use the Terraform icon or a folder icon or a git icon.

File: internal/terraform/datadir/module_types.go (hunk 12 above)
Comment (radeksimko) [reply internal/terraform/datadir/module_types.go:28]: Yep, agreed. 👍🏻

File: internal/langserver/handlers/command/module_calls.go [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+			moduleName = v[0]
+			subModuleName = v[1]
+		}
+
+		t := manifest.GetModuleType()
+
+		// build what we know
+		moduleInfo := moduleCall{
+			Name:             moduleName,
+			Source:           strings.TrimSuffix(manifest.SourceAddr, "/"),
```
[the remark this answers is not in the document]
Comment (jpogran) [reply internal/langserver/handlers/command/module_calls.go]: When I was making the treeitem description `sourceAdde@version` the trailing slash didn't match. This is not longer needed as we are not displaying that information that way any longer, as well as a display-centric bit of logic should not be in the LS

File: internal/terraform/datadir/module_types.go [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+const (
+	NOTFOUND   ModuleType = "notfound"
+	TFREGISTRY ModuleType = "tfregistry"
+	LOCAL      ModuleType = "local"
+	GITHUB     ModuleType = "github"
+	GITLAB     ModuleType = "gitlab"
+)
+
+func (r *ModuleRecord) GetModuleType() ModuleType {
+	if _, err := tfregistry.ParseRawProviderSourceString(r.SourceAddr); err == nil {
```
[the remark this answers is not in the document]
Comment (jpogran) [reply internal/terraform/datadir/module_types.go:29]: After discussion, we are going to keep the code as is and ticket future work at building a parser for Terraform module source addresses. Building such a parser is out of scope for this set of work and while technically incorrect `ParseRawProviderSourceString` works well enough to figure out if we are using the Terraform registry.

File: internal/langserver/handlers/command/module_calls_test.go [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+					SourceType:       "github",
+					DocsLink:         "",
+					DependentModules: []moduleCall{},
+				},
+			},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if got := parseModuleRecords(tt.records); !reflect.DeepEqual(got, tt.want) {
```
Comment (radeksimko) [inline internal/langserver/handlers/command/module_calls_test.go:84]: nitpick: [`cmp.Diff`](https://github.com/google/go-cmp) provides a better (more human-readable) way of comparing complex structures in tests and we already use it in many other tests, but it's certainly not a blocker for merging the PR!

File: internal/terraform/datadir/module_types.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (radeksimko) [reply internal/terraform/datadir/module_types.go:29]: Thank you for capturing all the useful context in that new comment! 👍🏻

**Review by radeksimko — APPROVED** [review]
🚢 🚀

File: internal/langserver/handlers/command/module_calls_test.go (hunk 16 above)
Comment (jpogran) [reply internal/langserver/handlers/command/module_calls_test.go]: omg that is so much nicer!
